File: vigenere.c
//Encrypt string with Vigenere's encryption by the specified key.
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define LENALPHA 26

int main(int argc, string argv[])
{
    //Check only one argument has been entered
    if (argc != 2)
    {
        printf("You have entered %i arguments, please enter only one.\n",  argc - 1);
        return 1;
    }
    //Check argument is valid word
    for (int i = 0, n = strlen(argv[1]) ; i < n ; i++)
    {
        if (!isalpha(argv[1][i]))
        {
            printf("You have entered %s, please enter an alphabetical word, with no digits or symbols.\n",  argv[1]);
            return 1;
        }

    }


    //Turn Key into array of sequence or integer keys.
    int key_len = strlen(argv[1]);
    int keys[key_len];
    char char_case;
    for (int i = 0; i < key_len ; i++)
    {
        char_case = isupper(argv[1][i]) ? 'A' : 'a';
        keys[i] = argv[1][i] - char_case;
    }

    //Get users text to encrypt.
    string plaintext = get_string("plaintext: ");
    string ciphertext = plaintext;

    //Apply Vigenere's encryption to each character that is a letter
    char c;
    for (int i = 0, j = 0, n = strlen(plaintext); i < n ; i++)
    {
        c = plaintext[i];

        //Check letter and case
        if isalpha(c)
        {
            j == key_len ? j %= key_len : j;
            char initial_letter = isupper(c) ? 'A' : 'a';

            //Apply Encryption
            c -= initial_letter;
            c += keys[j];
            c %= LENALPHA;
            c += initial_letter;

            j++;
        }

        ciphertext[i] = c;

    }

    //Return ciphertext to screen
    printf("ciphertext: %s\n", ciphertext);
    return 0;
}

File: vigenere.py
"""Vigenère Homework Problem"""
from cs50 import get_string


# Define global varaibles
LENALPHA = 26


def main(argv):
    """Encrypt plaintext message with key given using Vigenère's cipher."""

    # Ensure correct number of arguements
    if len(argv) != 2:

        # Raise program usage error
        raise Exception("Usage: python3 vigenere.py key")

    # Ensure valid key
    if not argv[1].isalpha():

        # Raise key usage error
        raise Exception("Usage - key: must be a string of alphabetical characters")

    # Initiate list to hold alphabetical indexes
    key = []

    # Iterate over characters in specified key in lowercase
    for char in argv[1].lower():

        # Convert character to alphabetical index
        aIndex = ord(char) - ord('a')

        # Append to key list
        key.append(aIndex)

    # Retrieve plain text
    plain = get_string("plaintext: ")

    # Initiate ciphertext and key index varaibles
    cipher = ""
    kIndex = 0

    # Iterate over characters in plain
    for char in plain:

        # Ensure key index is within range
        kIndex %= len(key)

        # Check if char is a capital
        if char.isupper():

            # Convert character to alphabetical index
            aIndex = ord(char) - ord('A')

            # Apply Vigenere's Cipher
            aIndex = (aIndex + key[kIndex]) % LENALPHA

            # Convert alphabetical index back to character
            char = chr(aIndex + ord('A'))

            # Increment key index by one
            kIndex += 1

        # Check if char is a capital
        elif char.islower():

            # Convert character to alphabetical index
            aIndex = ord(char) - ord('a')

            # Apply Caesar's Cipher
            aIndex = (aIndex + key[kIndex]) % LENALPHA

            # Convert alphabetical index back to character
            char = chr(aIndex + ord('a'))

            # Increment key index by one
            kIndex += 1

        # Add char to ciphertext
        cipher += char

    # Print ciphertext to user
    print("ciphertext: " + cipher)


if __name__ == "__main__":
    # If program run get programs arguements
    from sys import argv

    # Call main with argv
    main(argv)
